File: app/src/main/java/net/veldor/flibustaloader/utils/SortHandler.java
package net.veldor.flibustaloader.utils;

import net.veldor.flibustaloader.App;
import net.veldor.flibustaloader.selections.Author;
import net.veldor.flibustaloader.selections.FoundedBook;
import net.veldor.flibustaloader.selections.FoundedSequence;
import net.veldor.flibustaloader.selections.Genre;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

public class SortHandler {
    public static void sortBooks(ArrayList<FoundedBook> books) {
        if (App.getInstance().mBookSortOption != -1 && books != null && books.size() > 0) {
            Collections.sort(books, (lhs, rhs) -> {
                // -1 - less than, 1 - greater than, 0 - equal, all inversed for descending
                switch (App.getInstance().mBookSortOption) {
                    case 1:
                        int size1 = 0;
                        int size2 = 0;
                        // сортирую по размеру
                        String size_1 = lhs.size.replaceAll("[^\\d]", "");
                        String size_2 = rhs.size.replaceAll("[^\\d]", "");
                        if (!size_1.isEmpty())
                            size1 = Integer.parseInt(size_1);
                        if (!size_2.isEmpty())
                            size2 = Integer.parseInt(size_2);
                        if (size1 == size2)
                            return 0;
                        return size1 < size2 ? 1 : -1;
                    case 2:
                        // сортирую по количеству загрузок
                        String quantity1 = lhs.downloadsCount;
                        String quantity2 = rhs.downloadsCount;
                        int downloads1 = 0;
                        int downloads2 = 0;
                        if (quantity1 != null && !quantity1.isEmpty()) {
                            downloads1 = Integer.parseInt(quantity1.replaceAll("[^\\d]", ""));
                        }
                        if (quantity2 != null && !quantity2.isEmpty()) {
                            downloads2 = Integer.parseInt(quantity2.replaceAll("[^\\d]", ""));
                        }
                        if (downloads1 == downloads2)
                            return 0;
                        return downloads1 < downloads2 ? 1 : -1;
                    case 3:
                        // сортировка по серии
                        if (lhs.sequenceComplex.isEmpty()) {
                            return 1;
                        }
                        if (rhs.sequenceComplex.isEmpty()) {
                            return -1;
                        }
                        if (lhs.sequenceComplex.equals(rhs.sequenceComplex))
                            return 0;
                        return lhs.sequenceComplex.compareTo(rhs.sequenceComplex) > 0 ? 1 : -1;
                    case 4:
                        // сортировка по серии
                        if (lhs.genreComplex == null || lhs.genreComplex.isEmpty()) {
                            return 1;
                        }
                        if (rhs.genreComplex == null || rhs.genreComplex.isEmpty()) {
                            return -1;
                        }
                        if(lhs.genreComplex.equals(rhs.genreComplex))
                            return 0;
                        return lhs.genreComplex.compareTo(rhs.genreComplex) > 0 ? 1 : -1;
                    case 5:
                        // сортировка по серии
                        if (lhs.author.isEmpty()) {
                            return 1;
                        }
                        if (rhs.author.isEmpty()) {
                            return -1;
                        }
                        if(lhs.author.equals(rhs.author))
                            return 0;
                        return lhs.author.compareTo(rhs.author) > 0 ? 1 : -1;
                    default:
                    case 0:
                        // сортирую по названию книги
                        if(lhs.name.equals(rhs.name))
                            return 0;
                        return lhs.name.compareTo(rhs.name) > 0 ? 1 : -1;
                }
            });
        }
    }

    public static void sortAuthors(ArrayList<Author> mAuthors) {
        // отсортирую результат
        if (App.getInstance().mAuthorSortOptions != -1 && mAuthors != null && mAuthors.size() > 0) {
            Collections.sort(mAuthors, (lhs, rhs) -> {
                // -1 - less than, 1 - greater than, 0 - equal, all inversed for descending
                switch (App.getInstance().mAuthorSortOptions) {
                    case 1:
                        if(lhs.name.equals(rhs.name))
                            return 0;
                        return lhs.name.compareTo(rhs.name) > 0 ? -1 : 1;
                    case 2:
                        // сортирую по размеру
                        int size1 = Integer.parseInt(lhs.content.replaceAll("[^\\d]", ""));
                        int size2 = Integer.parseInt(rhs.content.replaceAll("[^\\d]", ""));
                        if(size1 == size2)
                            return 0;
                        return size1 < size2 ? 1 : -1;
                    case 3:
                        // сортирую по размеру
                        size1 = Integer.parseInt(lhs.content.replaceAll("[^\\d]", ""));
                        size2 = Integer.parseInt(rhs.content.replaceAll("[^\\d]", ""));
                        if(size1 == size2)
                            return 0;
                        return size1 < size2 ? -1 : 1;
                    default:
                    case 0:
                        // сортирую по названию книги
                        if((lhs).name.equals(rhs.name))
                            return 0;
                        return lhs.name.compareTo(rhs.name);
                }
            });
        }
    }

    public static void sortGenres(ArrayList<Genre> genres) {
        // отсортирую результат
        if (App.getInstance().mOtherSortOptions != -1 && genres != null && genres.size() > 0) {

            Collections.sort(genres, (lhs, rhs) -> {
                // -1 - less than, 1 - greater than, 0 - equal, all inversed for descending
                switch (App.getInstance().mOtherSortOptions) {
                    case 1:
                        if(lhs.label.equals(rhs.label))
                            return 0;
                        return lhs.label.compareTo(rhs.label) > 0 ? -1 : 1;
                    case 0:
                    default:
                        if(lhs.label.equals(rhs.label))
                            return 0;
                        // сортирую по названию книги
                        return lhs.label.compareTo(rhs.label);
                }
            });
        }
    }

    public static void sortSequences(ArrayList<FoundedSequence> sequences) {
        // отсортирую результат
        if (App.getInstance().mOtherSortOptions != -1 && sequences != null && sequences.size() > 0) {
            Collections.sort(sequences, (lhs, rhs) -> {
                // -1 - less than, 1 - greater than, 0 - equal, all inversed for descending
                switch (App.getInstance().mOtherSortOptions) {
                    case 1:
                        if(lhs.title.equals(rhs.title))
                            return 0;
                        return lhs.title.compareTo(rhs.title) > 0 ? -1 : 1;
                    case 0:
                    default:
                        if(lhs.title.equals(rhs.title))
                            return 0;
                        // сортирую по названию книги
                        return lhs.title.compareTo(rhs.title);
                }
            });
        }
    }
}

File: app/src/main/java/net/veldor/flibustaloader/utils/SortHandler.kt
package net.veldor.flibustaloader.utils

import android.os.Build
import androidx.core.text.isDigitsOnly
import net.veldor.flibustaloader.selections.*
import java.util.*

object SortHandler {
    @kotlin.jvm.JvmStatic
    fun sortBooks(books: ArrayList<FoundedEntity>, sortOption: Int, lastSortOption: Int) {
        books.sortWith(Comparator sort@{ lhs: FoundedEntity, rhs: FoundedEntity ->
            when (sortOption) {
                1 -> {
                    var size1 = 0
                    var size2 = 0
                    // сортирую по размеру
                    try {
                        val size11 = lhs.size!!.replace("[^\\d]".toRegex(), "")
                        if (size11.isNotEmpty() && size11.isDigitsOnly()) size1 = size11.toInt()
                    } catch (_: Exception) {
                    }
                    try {
                        val size21 = rhs.size!!.replace("[^\\d]".toRegex(), "")
                        if (size21.isNotEmpty() && size21.isDigitsOnly()) size2 = size21.toInt()
                    } catch (_: Exception) {
                    }
                    if (size1 == size2) return@sort 0
                    if (lastSortOption == sortOption) {
                        // invert
                        return@sort if (size1 < size2) -1 else 1
                    }
                    return@sort if (size1 < size2) 1 else -1
                }
                2 -> {
                    // сортирую по количеству загрузок
                    val quantity1 = lhs.downloadsCount
                    val quantity2 = rhs.downloadsCount
                    var downloads1 = 0
                    var downloads2 = 0
                    if (quantity1 != null && quantity1.isNotEmpty()) {
                        val first = quantity1.replace("[^\\d]".toRegex(), "")
                        if (first.isDigitsOnly()) downloads1 = first.toInt()
                    }
                    if (quantity2 != null && quantity2.isNotEmpty()) {
                        val second = quantity2.replace("[^\\d]".toRegex(), "")
                        if (second.isDigitsOnly()) downloads2 = second.toInt()
                    }
                    if (downloads1 == downloads2) return@sort 0
                    if (lastSortOption == sortOption) {
                        // invert
                        return@sort if (downloads1 < downloads2) -1 else 1
                    }
                    return@sort if (downloads1 < downloads2) 1 else -1
                }
                3 -> {
                    if (lhs.sequencesComplex == rhs.sequencesComplex) return@sort 0
                    // сортировка по серии
                    if (lhs.sequencesComplex.isEmpty()) {
                        return@sort 1
                    }
                    if (rhs.sequencesComplex.isEmpty()) {
                        return@sort -1
                    }
                    if (lastSortOption == sortOption) {
                        // invert
                        return@sort if (lhs.sequencesComplex < rhs.sequencesComplex) 1 else -1
                    }
                    return@sort if (lhs.sequencesComplex > rhs.sequencesComplex) 1 else -1
                }
                4 -> {
                    // сортировка по серии
                    if (lhs.genreComplex == null || lhs.genreComplex!!.isEmpty()) {
                        return@sort 1
                    }
                    if (rhs.genreComplex == null || rhs.genreComplex!!.isEmpty()) {
                        return@sort -1
                    }
                    if (lhs.genreComplex == rhs.genreComplex) return@sort 0
                    if (lastSortOption == sortOption) {
                        // invert
                        return@sort if (lhs.genreComplex!! < rhs.genreComplex!!) 1 else -1
                    }
                    return@sort if (lhs.genreComplex!! > rhs.genreComplex!!) 1 else -1
                }
                5 -> {
                    // сортировка по серии
                    if (lhs.author!!.isEmpty()) {
                        return@sort 1
                    }
                    if (rhs.author!!.isEmpty()) {
                        return@sort -1
                    }
                    if (lhs.author == rhs.author) return@sort 0
                    if (lastSortOption == sortOption) {
                        // invert
                        return@sort if (lhs.author!! < rhs.author!!) 1 else -1
                    }
                    return@sort if (lhs.author!! > rhs.author!!) 1 else -1
                }
                0 -> {
                    // сортирую по названию книги
                    if (lhs.name == rhs.name) return@sort 0
                    if (lastSortOption == sortOption) {
                        // invert
                        return@sort if (lhs.name!! < rhs.name!!) 1 else -1
                    }
                    return@sort if (lhs.name!! > rhs.name!!) 1 else -1
                }
                6 -> {
                    // сортирую по названию книги
                    if (lhs.downloaded == rhs.downloaded) return@sort 0
                    if (lastSortOption == sortOption) {
                        // invert
                        return@sort if (lhs.downloaded) -1 else 1
                    }
                    return@sort if (lhs.downloaded) 1 else -1
                }
                7 -> {
                    if (lhs.read == rhs.read) return@sort 0
                    if (lastSortOption == sortOption) {
                        // invert
                        return@sort if (lhs.read) -1 else 1
                    }
                    return@sort if (lhs.read) 1 else -1
                }
                8 -> {
                    if (lhs.format == rhs.format) return@sort 0
                    if (lhs.format.isNullOrEmpty()) {
                        return@sort 1
                    }
                    if (rhs.format.isNullOrEmpty()) {
                        return@sort -1
                    }
                    if (lastSortOption == sortOption) {
                        // invert
                        return@sort if (lhs.format!! < rhs.format!!) 1 else -1
                    }
                    return@sort if (lhs.format!! > rhs.format!!) 1 else -1
                }
                else -> {
                    if (lhs.name == rhs.name) return@sort 0
                    if (lastSortOption == sortOption) {
                        // invert
                        return@sort if (lhs.name!! > rhs.name!!) -1 else 1
                    }
                    return@sort if (lhs.name!! > rhs.name!!) 1 else -1
                }
            }
        })
    }

    @kotlin.jvm.JvmStatic
    fun sortAuthors(mAuthors: ArrayList<FoundedEntity>, sortOption: Int) {
        // отсортирую результат
        mAuthors.sortWith(Comparator sort@{ lhs: FoundedEntity, rhs: FoundedEntity ->
            when (sortOption) {
                1 -> {
                    if (lhs.name == rhs.name) return@sort 0
                    return@sort if (lhs.name!! > rhs.name!!) -1 else 1
                }
                2 -> {
                    // сортирую по размеру
                    val first = lhs.content.replace("[^\\d]".toRegex(), "")
                    val second = rhs.content.replace("[^\\d]".toRegex(), "")
                    val size1: Int = if (first == "") {
                        0
                    } else {
                        if (first.isDigitsOnly()) first.toInt() else 0
                    }
                    val size2: Int = if (second == "") {
                        0
                    } else {
                        if (second.isDigitsOnly()) second.toInt() else 0
                    }
                    if (size1 == size2) return@sort 0
                    return@sort if (size1 < size2) 1 else -1
                }
                3 -> {
                    // сортирую по размеру
                    val first = lhs.content.replace("[^\\d]".toRegex(), "")
                    val second = rhs.content.replace("[^\\d]".toRegex(), "")
                    val size1: Int = if (first == "") {
                        0
                    } else {
                        if (first.isDigitsOnly()) first.toInt() else 0
                    }
                    val size2: Int = if (second == "") {
                        0
                    } else {
                        if (second.isDigitsOnly()) second.toInt() else 0
                    }
                    if (size1 == size2) return@sort 0
                    return@sort if (size1 < size2) -1 else 1
                }
                0 -> {
                    // сортирую по названию книги
                    if (lhs.name == rhs.name) return@sort 0
                    return@sort lhs.name!!.compareTo(rhs.name!!)
                }
                else -> {
                    if (lhs.name == rhs.name) return@sort 0
                    return@sort lhs.name!!.compareTo(rhs.name!!)
                }
            }
        })
    }

    @kotlin.jvm.JvmStatic
    fun sortList(list: ArrayList<FoundedEntity>, sortOption: Int) {
        // отсортирую результат
        list.sortWith sort@{ lhs: FoundedEntity, rhs: FoundedEntity ->
            when (sortOption) {
                1 -> {
                    if (lhs.name == rhs.name) return@sort 0
                    return@sort if (lhs.name!! > rhs.name!!) -1 else 1
                }
                0 -> {
                    if (lhs.name == rhs.name) return@sort 0
                    // сортирую по названию книги
                    return@sort lhs.name!!.compareTo(rhs.name!!)
                }
                else -> {
                    if (lhs.name == rhs.name) return@sort 0
                    return@sort lhs.name!!.compareTo(rhs.name!!)
                }
            }
        }
    }


    @kotlin.jvm.JvmStatic
    fun sortLoadedBooks(mItems: ArrayList<Book>, which: Int, isInvert: Boolean) {
        mItems.sortWith sort@{ lhs: Book, rhs: Book ->
            var firstBook = lhs
            var secondBook = rhs
            if (isInvert) {
                // поменяю переменные местами
                firstBook = rhs
                secondBook = lhs
            }
            when (which) {
                0 ->                     // сортирую по названию книги
                    return@sort firstBook.name!!.compareTo(secondBook.name!!)
                1 -> {
                    // сортирую по размеру
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                        return@sort if (firstBook.file!!.length() < secondBook.file!!.length()) 1 else -1
                    }
                    return@sort if (firstBook.fileCompat!!.length() < secondBook.fileCompat!!.length()) 1 else -1
                }
                3 ->                     // отсортирую по формату
                    return@sort firstBook.extension!!.compareTo(secondBook.extension!!)
                4 ->                     // отсортирую по формату
                    return@sort firstBook.author!!.compareTo(secondBook.author!!)
                5 -> {
                    // отсортирую по времени добавления
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                        return@sort if (firstBook.file!!.lastModified() < secondBook.file!!.lastModified()) 1 else -1
                    }
                    return@sort if (firstBook.fileCompat!!.lastModified() < secondBook.fileCompat!!.lastModified()) 1 else -1
                }
            }
            0
        }
    }
}